clippy_lints/src/assertions_on_constants.rs

```diff
@@ -63,7 +65,93 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for AssertionsOnConstants {
                 return;
             }
-            lint_assert_cb(false);
+            if let Some((panic_message, is_true)) = assert_with_message(&cx, e) {
+                if is_true {
+                    span_help_and_lint(
+                        cx,
+                        ASSERTIONS_ON_CONSTANTS,
+                        e.span,
+                        "`assert!(true)` will be optimized out by the compiler",
+                        "remove it",
+                    );
+                } else if panic_message.starts_with("assertion failed: ") {
+                    span_help_and_lint(
+                        cx,
+                        ASSERTIONS_ON_CONSTANTS,
+                        e.span,
+                        "`assert!(false)` should probably be replaced",
+                        "use `panic!()` or `unreachable!()`",
+                    );
+                } else {
+                    span_help_and_lint(
+                        cx,
+                        ASSERTIONS_ON_CONSTANTS,
+                        e.span,
+                        &format!("`assert!(false, \"{}\")` should probably be replaced", panic_message,),
+                        &format!(
+                            "use `panic!(\"{}\")` or `unreachable!(\"{}\")`",
+                            panic_message, panic_message,
+                        ),
+                    );
+                }
+            }
+        }
+    }
+}
+
+// fn get_assert_args(snip: String) -> Option<Vec<String>> {
+//
+// }
+
+fn assert_with_message<'a, 'tcx>(cx: &LateContext<'a, 'tcx>, expr: &'tcx Expr) -> Option<(LocalInternedString, bool)> {
+    if_chain! {
+        if let ExprKind::Match(ref expr, ref arms, MatchSource::IfDesugar { contains_else_clause: false }) = expr.kind;
+        // match expr
+        if let ExprKind::DropTemps(ref expr) = expr.kind;
+        if let ExprKind::Unary(UnOp::UnNot, ref expr) = expr.kind;
+        //if let ExprKind::Lit(ref lit) = expr.kind;
+        if let Some((Constant::Bool(is_true), _)) = constant(cx, cx.tables, expr);
+        //if is_true;
+        // match arms
+        // arm 1 pattern
+        if let PatKind::Lit(ref lit_expr) = arms[0].pat.kind;
+        if let ExprKind::Lit(ref lit) = lit_expr.kind;
+        if let LitKind::Bool(true) = lit.node;
+        //if let LitKind::Bool(true) = lit1.node;
+        // arm 1 block
+        if let ExprKind::Block(ref block1, _) = arms[0].body.kind;
+        if let Some(trailing_expr1) = &block1.expr;
+        if block1.stmts.len() == 0;
+        //
+        if let ExprKind::Block(ref actual_block1, _) = trailing_expr1.kind;
+        if let Some(block1_expr) = &actual_block1.expr;
+        // function call
+        if let ExprKind::Call(ref func, ref args) = block1_expr.kind;
+        if let ExprKind::Path(ref path) = func.kind;
+        // ["{{root}}", "std", "rt", "begin_panic"] does not work
+        if match_qpath(path, &["$crate", "rt", "begin_panic"]);
```

Comment:
There's already a path for `begin_panic` in `paths`:
https://github.com/rust-lang/rust-clippy/blob/54bf4ffd626970e831bb80c037f804a3b3450835/clippy_lints/src/utils/paths.rs#L8

It is mostly used with `match_def_path`. Maybe you can do this here also?
https://github.com/rust-lang/rust-clippy/blob/edd90473ec5ba29b9ae1ee706c982e7046a6e63e/clippy_lints/src/panic_unimplemented.rs#L52-L55

Reply:
I also added a helper function

```rust
fn match_function_call<'a, 'tcx>(cx: &LateContext<'a, 'tcx>, expr: &'tcx Expr, path: &[&str]) -> Option<&'a [Expr]> {}
```

that matches the path and returns the arguments.

Maybe this would be worth adding to the utils.